common/djangoapps/track/transformers.py

```diff
@@ -0,0 +1,502 @@
+
+            An *exact transformer* will only handle events whose name matches
+            name of the transformer exactly.
+
+            Exact transformers always take precedence over prefix transformers.
+
+            Transformers without a name will not be added to the registry, and
+            cannot be accessed via the `EventTransformer.create_transformer()`
+            classmethod.
+
+        is_legacy_event:
+            If an event is a legacy event, it needs to set event_type to the
+            legacy name for the event, and may need to set certain event fields
+            to maintain backward compatiblity.  If an event needs to provide
+            legacy support in some contexts, `is_legacy_event` can be defined
+            as a property to add dynamic behavior.
+
+            Default: False
+
+        legacy_event_type:
+            If the event is or can be a legacy event, it should define
+            the legacy value for the event_type field here.
+
+    Processing methods.  Override these to provide the behavior needed for your
+    particular EventTransformer:
+
+        self.process_legacy_fields():
+            This method should modify the event payload in any way necessary to
+            support legacy event types.  It will only be run if
+            `is_legacy_event` returns a True value.
+
+        self.process_event()
+            This method modifies the event payload unconditionally.  It will
+            always be run.
+    """
+    def __init__(self, *args, **kwargs):
+        super(EventTransformer, self).__init__(*args, **kwargs)
+        self.load_payload()
+
+    # Properties to be overridden
+
+    is_legacy_event = False
+
+    @property
+    def legacy_event_type(self):
+        """
+        Override this as an attribute or property to provide the value for
+        the event's `event_type`, if it does not match the event's `name`.
+        """
+        raise NotImplementedError
+
+    # Convenience properties
+
+    @property
+    def name(self):
+        """
+        Returns the event's name.
+        """
+        return self[u'name']
+
+    @property
+    def context(self):
+        """
+        Returns the event's context dict.
+        """
+        return self.get(u'context', {})
+
+    # Transform methods
+
+    def load_payload(self):
+        """
+        Create a data version of self[u'event'] at self.event
+        """
+        if u'event' in self:
+            if isinstance(self[u'event'], basestring):
+                self.event = json.loads(self[u'event'])
+            else:
+                self.event = self[u'event']
+
+    def dump_payload(self):
+        """
+        Write self.event back to self[u'event'].
+
+        Keep the same format we were originally given.
+        """
+        if isinstance(self.get(u'event'), basestring):
+            self[u'event'] = json.dumps(self.event)
+        else:
+            self[u'event'] = self.event
```

Comment:
I would recommend we don't try to change this at the moment. It requires a bunch of coordination with external people.